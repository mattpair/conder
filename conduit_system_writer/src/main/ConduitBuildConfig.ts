

export type ConduitBuildConfig = {
    project: string;
    dependents?: {
        [path in string]: {
            language: "typescript";
        };
    };
};
