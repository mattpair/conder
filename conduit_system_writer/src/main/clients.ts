import { CompiledTypes, Utilities } from 'conduit_compiler';

export const a: string = `${12}`

function typeToTS(t: CompiledTypes.RealType):string {
    return `${t.val.name}${t.isArray ? "[]" : ""}`
}

export function generateClients(url: string, manifest: CompiledTypes.Manifest, models: string[]) {
    const clients: string[] = []
    manifest.inScope.forEach(fn => {
        if (fn.kind !== "Function") {
            return
        }
        const param = fn.parameter.differentiate()
        const ret = fn.returnType

        let returnType = ''
        let followOn = ''
        let body =  ''
        let paramString = ''
        let beforeReq = ''

        switch(param.kind) {
            case "UnaryParameter":
                paramString = `a: ${typeToTS(param.type)}`
                beforeReq = 'const body = JSON.stringify(a)'
                body = `
                body,
                headers: {
                    "content-type": "application/json",
                    "content-length": \`\${body.length}\`
                },
                `
                break;
            case "NoParameter":
                break;

            default: Utilities.assertNever(param)
        }
        switch (ret.kind) {
            case "real type":
                returnType = `: Promise<${typeToTS(ret)}>`
                followOn = '.then( data=> data.json())'
                break;
            case "VoidReturnType":
                break;

            default: Utilities.assertNever(ret)
        }

        clients.push(`
        export function ${fn.name}(${paramString})${returnType} {
            ${beforeReq}
            return fetch("${url}/${fn.name}", {
                ${body}
                method: "${fn.method}"
            })${followOn}

        }`) 
    })
    return  `
        const url = '${url}'

        ${models.join("\n")}

        export function hello() {
            return fetch(url)
        }

        ${clients.join("\n")}
        `
}