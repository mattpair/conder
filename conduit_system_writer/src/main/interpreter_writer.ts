import { CompiledTypes } from "conduit_compiler";


export function writeOperationInterpreter(manifest: CompiledTypes.Manifest): string {
    const op_match: string[] = []
    const op_type: string[] = []
    const runtime_ent: string[] = [
        "None",
        "Err(String)"
    ]

    manifest.supportedOperations.forEach(op => {
        switch(op.kind) {
            case "insert": {
                const store = manifest.inScope.getEntityOfType(op.storeName, "HierarchicalStore")
                const op_name = `Insert_${op.storeName}`
                op_type.push(`${op_name}(${store.typeName})`)

                op_match.push(`
                Op::${op_name}(v) => {
                    return match insert_${op.storeName}(&client, v).await {
                        Ok(()) => OpResult::None,
                        Err(err) => OpResult::Err(err.to_string())
                    };
                }`)
                break
            }
            case "noop":
                op_type.push(`Noop`)
                op_match.push(`Op::Noop => OpResult::None`)
                break
            case "query": {
                const store = manifest.inScope.getEntityOfType(op.storeName, "HierarchicalStore")
                const op_name = `Query_${op.storeName}`
                op_type.push(`${op_name}(${store.specName})`)
                const result_name = `QueryResult${op.storeName}`
                runtime_ent.push(`${result_name}(Vec<${store.typeName}>)`)
                op_match.push(`
                Op::${op_name}(v) => {
                    return match query_interpreter_${store.name}(v, &client).await {
                        Ok(out) => OpResult::${result_name}(out),
                        Err(err) => OpResult::Err(err.to_string())
                    };
                }`)
                break
            }
            case "return input":
                break
        }
    })

    return `
    #[derive(Serialize, Deserialize, Clone)]
    #[serde(tag = "kind")]
    enum Op {
        ${op_type.join(",\n")}
    }

    #[derive(Serialize, Deserialize, Clone)]
    #[serde(tag = "kind")]
    enum RuntimeEnt {
        ${runtime_ent.join(",\n")}
    }

    async fn op_interpreter(op: Op, client: &Client) -> RuntimeEnt {
        return match op {
            ${op_match.join(",\n")}
        };
    }
        
    `
}