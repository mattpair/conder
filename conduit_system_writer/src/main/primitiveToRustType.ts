import { CompiledTypes, Lexicon, Utilities } from 'conduit_parser';
export function primitiveToRustType(type: Lexicon.PrimitiveUnion): string {
    switch (type) {
        case Lexicon.Symbol.double:
            return "f64";

        case Lexicon.Symbol.float:
            return "f32";

        case Lexicon.Symbol.int32:
            return "i32";

        case Lexicon.Symbol.int64:
            return "i64";

        case Lexicon.Symbol.string:
            return "String";

        case Lexicon.Symbol.uint32:
            return "i32";

        case Lexicon.Symbol.uint64:
            return "i64";

        case Lexicon.Symbol.bool:
            return "bool";


        case Lexicon.Symbol.bytes:
            return "Vec<u8>";

        default: Utilities.assertNever(type);
    }
}
