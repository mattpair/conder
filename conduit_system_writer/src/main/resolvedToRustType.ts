import { CompiledTypes } from 'conduit_parser';
import { assertNever } from 'conduit_parser/dist/src/main/utils';
import { primitiveToRustType } from './primitiveToRustType';

export function resolvedToRustType(r: CompiledTypes.ResolvedType, inScope: CompiledTypes.ScopeMap): string {
    let base = '';
    switch (r.kind) {
        case "Primitive":
            base = primitiveToRustType(r.val);
            break;

        case "CustomType":
            const ent = inScope.getEntityOfType(r.type, "Enum", "Struct");
            switch (ent.kind) {
                case "Struct":
                    base = r.type;
                    break;

                case "Enum":
                    base = 'i64';
                    break;
            }
            break;

        default: assertNever(r);
    }
    let prefix = '';
    let suffix = '';
    switch (r.modification) {
        case "array":
            prefix = 'Vec<';
            suffix = '>';
        case "none":
            break;

        case "optional":
            prefix = 'Option<';
            suffix = '>';
            break;

        default: assertNever(r.modification);
    }
    return `${prefix}${base}${suffix}`;
}
