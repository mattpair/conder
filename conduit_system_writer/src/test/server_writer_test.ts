import { writeRustAndContainerCode } from "../main/server_writer"
import { CompiledTypes, Lexicon, compileFiles } from "conduit_compiler"

function TestCodeGen(description: string, conduit: string) {
    test(description, async () => {
        
        const r = await writeRustAndContainerCode.func({manifest: compileFiles({test: () => conduit})})
        const mainFiles = r.backend.main.files.filter(f => !(/Cargo/.test(f.name)))
        expect(mainFiles).toMatchSnapshot("main files")
        expect(r.backend.postgres.files).toMatchSnapshot("postgres files")
    })
}

TestCodeGen("only one struct", `
struct test {
    onlyField: bool 
}
`)

TestCodeGen("struct containing struct stored", `
struct inner {
    fieldA: double 
}

struct outer {
    inner: inner
}

outerStore: Array outer = []

`)

TestCodeGen("store containing struct containing array", `
struct inner {
    fieldA: int32 
}

struct outer {
    inners: Array inner
}

outerStore: Array outer = []

`)

TestCodeGen("inserting simple struct", `
struct simple {
    data:double 
}

simpleStore: Array simple = []

function insert(s: simple) {
    simpleStore.append(s)
}
`)

TestCodeGen("inserting struct containing struct", `
struct simple {
    data: bool 
}

struct wrapper {
    innard: simple 
}

wrapStore: Array wrapper = []

function insert(s: wrapper) {
    wrapStore.append(s)
}
`)

TestCodeGen("inserting struct containing struct containing primitive array", `
struct simple {
    data: Array string
}

struct wrapper {
    innard: simple 
}

wrapStore: Array wrapper = []

function insert(s: wrapper) {
    wrapStore.append(s)
}
`)

TestCodeGen("inserting struct containing struct array", `
struct simple {
    data: string 
}

struct wrapper {
    innard: Array simple
}

wrapStore: Array wrapper = []

function insert(s: wrapper) {
    wrapStore.append(s)
}
`)