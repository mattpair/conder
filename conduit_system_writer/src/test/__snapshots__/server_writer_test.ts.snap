// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inserting simple struct: main files 1`] = `
Array [
  Object {
    "content": "
                            #![allow(non_snake_case)]
                            #![allow(non_camel_case_types)]
                            #![allow(redundant_semicolon)]
                            use tokio_postgres::{NoTls, Client};
                            use actix_web::{web, App, HttpResponse, HttpServer, Responder};
                            use std::env;
                            use serde::{Deserialize, Serialize};
                            use tokio_postgres::error::{Error};
                
                
                            struct AppData {
                                client: Client
                            }
                            
                            #[derive(Serialize, Deserialize)]
                            struct City {
                                name: String,
                                location: i32
                            }
                            
                    
                            
                        #[derive(Serialize, Deserialize)]
                        struct simple {
                            data: f64
                        }
                    
                    
                            
        async fn internal_insert(s: simple, client: &Client)  -> Result<(), Error> {
            client.query(\\"insert into simpleStore(data) values ($1)\\", &[&s.data]).await?;
return Ok(())
        }

        async fn external_insert(input: web::Json<simple>, data: web::Data<AppData>) -> impl Responder {
            let s = input.into_inner();
let client = &data.client;
            return match internal_insert(s,client).await {
                    Ok(()) => HttpResponse::Ok().finish(),
                    Err(err) => {
                        HttpResponse::BadRequest().body(format!(\\"Failure caused by: {}\\", err))
                    }
                };
                
        }
                
        
                    
                            #[actix_rt::main]
                            async fn main() -> std::io::Result<()> {
                                HttpServer::new(|| {
                                    App::new()
                                        .data_factory(|| make_app_data())
                                        .route(\\"/\\", web::get().to(index))
                                        .route(\\"/insert\\", web::post().to(external_insert))
                                })
                                .bind(\\"0.0.0.0:8080\\")?
                                .run()
                                .await
                            }
                    
                            async fn index(data: web::Data<AppData>) -> impl Responder {
                                let mut rows = match data.client.query(\\"select name, location from cities\\", &[]).await {
                                    Ok(rows) => rows,
                                    Err(err) => panic!(\\"didn't succeed: {}\\", err)
                                };
                            
                                let mut out = Vec::with_capacity(rows.len());
                            
                                while let Some(row) = rows.pop() {
                                    out.push(City {
                                        name: row.get(0),
                                        location: row.get(1)
                                    })
                                }
                                return HttpResponse::Ok().json(out);
                            }
                
                            async fn make_app_data() -> Result<AppData, ()> {
                                let host = match env::var(\\"POSTGRES_SERVICE_HOST\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres location: {}\\", e)
                                };
                                let pwd = match env::var(\\"POSTGRES_PASSWORD\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres password: {}\\", e)
                                };
                            
                                let (client, connection) = match tokio_postgres::connect(&format!(\\"host={} user=postgres password={}\\", host, pwd), NoTls).await {
                                    Ok(out) => out,
                                    Err(e) => panic!(\\"couldn't create connection: {}\\", e)
                                };
                                
                                // The connection object performs the actual communication with the database,
                                // so spawn it off to run on its own.
                                actix_rt::spawn(async move {
                                    if let Err(e) = connection.await {
                                        eprintln!(\\"connection error: {}\\", e);
                                    }
                                });
                                
                                
                                return Ok(AppData {
                                    client: client,
                                });
                            }
                            
                        ",
    "name": ".deploy/main/src/main.rs",
  },
]
`;

exports[`inserting simple struct: postgres files 1`] = `
Array [
  Object {
    "content": "

                        CREATE TABLE cities (
                            name            varchar(80),
                            location        int
                        );
                
                        
        CREATE TABLE simpleStore (
            data	double precision,
conduit_entity_id	INT	GENERATED ALWAYS AS ENTITY ID,
PRIMARY KEY(conduit_entity_id)
        );
                        
                        
                        insert into cities(name, location)
                        values ('detroit', 12)",
    "name": ".deploy/postgres/startup/init.sql",
  },
]
`;

exports[`inserting struct containing struct containing primitive array: main files 1`] = `
Array [
  Object {
    "content": "
                            #![allow(non_snake_case)]
                            #![allow(non_camel_case_types)]
                            #![allow(redundant_semicolon)]
                            use tokio_postgres::{NoTls, Client};
                            use actix_web::{web, App, HttpResponse, HttpServer, Responder};
                            use std::env;
                            use serde::{Deserialize, Serialize};
                            use tokio_postgres::error::{Error};
                
                
                            struct AppData {
                                client: Client
                            }
                            
                            #[derive(Serialize, Deserialize)]
                            struct City {
                                name: String,
                                location: i32
                            }
                            
                    
                            
                        #[derive(Serialize, Deserialize)]
                        struct simple {
                            data: String
                        }
                    

                        #[derive(Serialize, Deserialize)]
                        struct wrapper {
                            innard: simple
                        }
                    
                    
                            
        async fn internal_insert(s: wrapper, client: &Client)  -> Result<(), Error> {
            let ret0 = client.query(\\"insert into wrapStore_innard(data) values ($1) RETURNING conduit_entity_id\\", &[&s.innard.data]).await?;
client.query(\\"insert into wrapStore(innard) values ($1)\\", &[&(ret0.get(0))]).await?;
return Ok(())
        }

        async fn external_insert(input: web::Json<wrapper>, data: web::Data<AppData>) -> impl Responder {
            let s = input.into_inner();
let client = &data.client;
            return match internal_insert(s,client).await {
                    Ok(()) => HttpResponse::Ok().finish(),
                    Err(err) => {
                        HttpResponse::BadRequest().body(format!(\\"Failure caused by: {}\\", err))
                    }
                };
                
        }
                
        
                    
                            #[actix_rt::main]
                            async fn main() -> std::io::Result<()> {
                                HttpServer::new(|| {
                                    App::new()
                                        .data_factory(|| make_app_data())
                                        .route(\\"/\\", web::get().to(index))
                                        .route(\\"/insert\\", web::post().to(external_insert))
                                })
                                .bind(\\"0.0.0.0:8080\\")?
                                .run()
                                .await
                            }
                    
                            async fn index(data: web::Data<AppData>) -> impl Responder {
                                let mut rows = match data.client.query(\\"select name, location from cities\\", &[]).await {
                                    Ok(rows) => rows,
                                    Err(err) => panic!(\\"didn't succeed: {}\\", err)
                                };
                            
                                let mut out = Vec::with_capacity(rows.len());
                            
                                while let Some(row) = rows.pop() {
                                    out.push(City {
                                        name: row.get(0),
                                        location: row.get(1)
                                    })
                                }
                                return HttpResponse::Ok().json(out);
                            }
                
                            async fn make_app_data() -> Result<AppData, ()> {
                                let host = match env::var(\\"POSTGRES_SERVICE_HOST\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres location: {}\\", e)
                                };
                                let pwd = match env::var(\\"POSTGRES_PASSWORD\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres password: {}\\", e)
                                };
                            
                                let (client, connection) = match tokio_postgres::connect(&format!(\\"host={} user=postgres password={}\\", host, pwd), NoTls).await {
                                    Ok(out) => out,
                                    Err(e) => panic!(\\"couldn't create connection: {}\\", e)
                                };
                                
                                // The connection object performs the actual communication with the database,
                                // so spawn it off to run on its own.
                                actix_rt::spawn(async move {
                                    if let Err(e) = connection.await {
                                        eprintln!(\\"connection error: {}\\", e);
                                    }
                                });
                                
                                
                                return Ok(AppData {
                                    client: client,
                                });
                            }
                            
                        ",
    "name": ".deploy/main/src/main.rs",
  },
]
`;

exports[`inserting struct containing struct containing primitive array: postgres files 1`] = `
Array [
  Object {
    "content": "

                        CREATE TABLE cities (
                            name            varchar(80),
                            location        int
                        );
                
                        
        CREATE TABLE wrapStore_innard (
            data	text[],
conduit_entity_id	INT	GENERATED ALWAYS AS ENTITY ID,
PRIMARY KEY(conduit_entity_id)
        );

        CREATE TABLE wrapStore (
            innard	INT,
conduit_entity_id	INT	GENERATED ALWAYS AS ENTITY ID,
constraint fk_innard
                                FOREIGN KEY(innard)
                                    REFERENCES wrapStore_innard(conduit_entity_id),
PRIMARY KEY(conduit_entity_id)
        );
                        
                        
                        insert into cities(name, location)
                        values ('detroit', 12)",
    "name": ".deploy/postgres/startup/init.sql",
  },
]
`;

exports[`inserting struct containing struct: main files 1`] = `
Array [
  Object {
    "content": "
                            #![allow(non_snake_case)]
                            #![allow(non_camel_case_types)]
                            #![allow(redundant_semicolon)]
                            use tokio_postgres::{NoTls, Client};
                            use actix_web::{web, App, HttpResponse, HttpServer, Responder};
                            use std::env;
                            use serde::{Deserialize, Serialize};
                            use tokio_postgres::error::{Error};
                
                
                            struct AppData {
                                client: Client
                            }
                            
                            #[derive(Serialize, Deserialize)]
                            struct City {
                                name: String,
                                location: i32
                            }
                            
                    
                            
                        #[derive(Serialize, Deserialize)]
                        struct simple {
                            data: bool
                        }
                    

                        #[derive(Serialize, Deserialize)]
                        struct wrapper {
                            innard: simple
                        }
                    
                    
                            
        async fn internal_insert(s: wrapper, client: &Client)  -> Result<(), Error> {
            let ret0 = client.query(\\"insert into wrapStore_innard(data) values ($1) RETURNING conduit_entity_id\\", &[&s.innard.data]).await?;
client.query(\\"insert into wrapStore(innard) values ($1)\\", &[&(ret0.get(0))]).await?;
return Ok(())
        }

        async fn external_insert(input: web::Json<wrapper>, data: web::Data<AppData>) -> impl Responder {
            let s = input.into_inner();
let client = &data.client;
            return match internal_insert(s,client).await {
                    Ok(()) => HttpResponse::Ok().finish(),
                    Err(err) => {
                        HttpResponse::BadRequest().body(format!(\\"Failure caused by: {}\\", err))
                    }
                };
                
        }
                
        
                    
                            #[actix_rt::main]
                            async fn main() -> std::io::Result<()> {
                                HttpServer::new(|| {
                                    App::new()
                                        .data_factory(|| make_app_data())
                                        .route(\\"/\\", web::get().to(index))
                                        .route(\\"/insert\\", web::post().to(external_insert))
                                })
                                .bind(\\"0.0.0.0:8080\\")?
                                .run()
                                .await
                            }
                    
                            async fn index(data: web::Data<AppData>) -> impl Responder {
                                let mut rows = match data.client.query(\\"select name, location from cities\\", &[]).await {
                                    Ok(rows) => rows,
                                    Err(err) => panic!(\\"didn't succeed: {}\\", err)
                                };
                            
                                let mut out = Vec::with_capacity(rows.len());
                            
                                while let Some(row) = rows.pop() {
                                    out.push(City {
                                        name: row.get(0),
                                        location: row.get(1)
                                    })
                                }
                                return HttpResponse::Ok().json(out);
                            }
                
                            async fn make_app_data() -> Result<AppData, ()> {
                                let host = match env::var(\\"POSTGRES_SERVICE_HOST\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres location: {}\\", e)
                                };
                                let pwd = match env::var(\\"POSTGRES_PASSWORD\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres password: {}\\", e)
                                };
                            
                                let (client, connection) = match tokio_postgres::connect(&format!(\\"host={} user=postgres password={}\\", host, pwd), NoTls).await {
                                    Ok(out) => out,
                                    Err(e) => panic!(\\"couldn't create connection: {}\\", e)
                                };
                                
                                // The connection object performs the actual communication with the database,
                                // so spawn it off to run on its own.
                                actix_rt::spawn(async move {
                                    if let Err(e) = connection.await {
                                        eprintln!(\\"connection error: {}\\", e);
                                    }
                                });
                                
                                
                                return Ok(AppData {
                                    client: client,
                                });
                            }
                            
                        ",
    "name": ".deploy/main/src/main.rs",
  },
]
`;

exports[`inserting struct containing struct: postgres files 1`] = `
Array [
  Object {
    "content": "

                        CREATE TABLE cities (
                            name            varchar(80),
                            location        int
                        );
                
                        
        CREATE TABLE wrapStore_innard (
            data	boolean,
conduit_entity_id	INT	GENERATED ALWAYS AS ENTITY ID,
PRIMARY KEY(conduit_entity_id)
        );

        CREATE TABLE wrapStore (
            innard	INT,
conduit_entity_id	INT	GENERATED ALWAYS AS ENTITY ID,
constraint fk_innard
                                FOREIGN KEY(innard)
                                    REFERENCES wrapStore_innard(conduit_entity_id),
PRIMARY KEY(conduit_entity_id)
        );
                        
                        
                        insert into cities(name, location)
                        values ('detroit', 12)",
    "name": ".deploy/postgres/startup/init.sql",
  },
]
`;

exports[`only one struct: main files 1`] = `
Array [
  Object {
    "content": "
                            #![allow(non_snake_case)]
                            #![allow(non_camel_case_types)]
                            #![allow(redundant_semicolon)]
                            use tokio_postgres::{NoTls, Client};
                            use actix_web::{web, App, HttpResponse, HttpServer, Responder};
                            use std::env;
                            use serde::{Deserialize, Serialize};
                            use tokio_postgres::error::{Error};
                
                
                            struct AppData {
                                client: Client
                            }
                            
                            #[derive(Serialize, Deserialize)]
                            struct City {
                                name: String,
                                location: i32
                            }
                            
                    
                            
                        #[derive(Serialize, Deserialize)]
                        struct test {
                            onlyField: bool
                        }
                    
                    
                            
                    
                            #[actix_rt::main]
                            async fn main() -> std::io::Result<()> {
                                HttpServer::new(|| {
                                    App::new()
                                        .data_factory(|| make_app_data())
                                        .route(\\"/\\", web::get().to(index))
                                        
                                })
                                .bind(\\"0.0.0.0:8080\\")?
                                .run()
                                .await
                            }
                    
                            async fn index(data: web::Data<AppData>) -> impl Responder {
                                let mut rows = match data.client.query(\\"select name, location from cities\\", &[]).await {
                                    Ok(rows) => rows,
                                    Err(err) => panic!(\\"didn't succeed: {}\\", err)
                                };
                            
                                let mut out = Vec::with_capacity(rows.len());
                            
                                while let Some(row) = rows.pop() {
                                    out.push(City {
                                        name: row.get(0),
                                        location: row.get(1)
                                    })
                                }
                                return HttpResponse::Ok().json(out);
                            }
                
                            async fn make_app_data() -> Result<AppData, ()> {
                                let host = match env::var(\\"POSTGRES_SERVICE_HOST\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres location: {}\\", e)
                                };
                                let pwd = match env::var(\\"POSTGRES_PASSWORD\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres password: {}\\", e)
                                };
                            
                                let (client, connection) = match tokio_postgres::connect(&format!(\\"host={} user=postgres password={}\\", host, pwd), NoTls).await {
                                    Ok(out) => out,
                                    Err(e) => panic!(\\"couldn't create connection: {}\\", e)
                                };
                                
                                // The connection object performs the actual communication with the database,
                                // so spawn it off to run on its own.
                                actix_rt::spawn(async move {
                                    if let Err(e) = connection.await {
                                        eprintln!(\\"connection error: {}\\", e);
                                    }
                                });
                                
                                
                                return Ok(AppData {
                                    client: client,
                                });
                            }
                            
                        ",
    "name": ".deploy/main/src/main.rs",
  },
]
`;

exports[`only one struct: postgres files 1`] = `
Array [
  Object {
    "content": "

                        CREATE TABLE cities (
                            name            varchar(80),
                            location        int
                        );
                
                        
                        
                        
                        insert into cities(name, location)
                        values ('detroit', 12)",
    "name": ".deploy/postgres/startup/init.sql",
  },
]
`;

exports[`store containing struct containing array: main files 1`] = `
Array [
  Object {
    "content": "
                            #![allow(non_snake_case)]
                            #![allow(non_camel_case_types)]
                            #![allow(redundant_semicolon)]
                            use tokio_postgres::{NoTls, Client};
                            use actix_web::{web, App, HttpResponse, HttpServer, Responder};
                            use std::env;
                            use serde::{Deserialize, Serialize};
                            use tokio_postgres::error::{Error};
                
                
                            struct AppData {
                                client: Client
                            }
                            
                            #[derive(Serialize, Deserialize)]
                            struct City {
                                name: String,
                                location: i32
                            }
                            
                    
                            
                        #[derive(Serialize, Deserialize)]
                        struct inner {
                            fieldA: i32
                        }
                    

                        #[derive(Serialize, Deserialize)]
                        struct outer {
                            inners: inner
                        }
                    
                    
                            
                    
                            #[actix_rt::main]
                            async fn main() -> std::io::Result<()> {
                                HttpServer::new(|| {
                                    App::new()
                                        .data_factory(|| make_app_data())
                                        .route(\\"/\\", web::get().to(index))
                                        
                                })
                                .bind(\\"0.0.0.0:8080\\")?
                                .run()
                                .await
                            }
                    
                            async fn index(data: web::Data<AppData>) -> impl Responder {
                                let mut rows = match data.client.query(\\"select name, location from cities\\", &[]).await {
                                    Ok(rows) => rows,
                                    Err(err) => panic!(\\"didn't succeed: {}\\", err)
                                };
                            
                                let mut out = Vec::with_capacity(rows.len());
                            
                                while let Some(row) = rows.pop() {
                                    out.push(City {
                                        name: row.get(0),
                                        location: row.get(1)
                                    })
                                }
                                return HttpResponse::Ok().json(out);
                            }
                
                            async fn make_app_data() -> Result<AppData, ()> {
                                let host = match env::var(\\"POSTGRES_SERVICE_HOST\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres location: {}\\", e)
                                };
                                let pwd = match env::var(\\"POSTGRES_PASSWORD\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres password: {}\\", e)
                                };
                            
                                let (client, connection) = match tokio_postgres::connect(&format!(\\"host={} user=postgres password={}\\", host, pwd), NoTls).await {
                                    Ok(out) => out,
                                    Err(e) => panic!(\\"couldn't create connection: {}\\", e)
                                };
                                
                                // The connection object performs the actual communication with the database,
                                // so spawn it off to run on its own.
                                actix_rt::spawn(async move {
                                    if let Err(e) = connection.await {
                                        eprintln!(\\"connection error: {}\\", e);
                                    }
                                });
                                
                                
                                return Ok(AppData {
                                    client: client,
                                });
                            }
                            
                        ",
    "name": ".deploy/main/src/main.rs",
  },
]
`;

exports[`store containing struct containing array: postgres files 1`] = `
Array [
  Object {
    "content": "

                        CREATE TABLE cities (
                            name            varchar(80),
                            location        int
                        );
                
                        
        CREATE TABLE outerStore_inners (
            fieldA	integer,
conduit_entity_id	INT	GENERATED ALWAYS AS ENTITY ID,
PRIMARY KEY(conduit_entity_id)
        );

        CREATE TABLE outerStore (
            conduit_entity_id	INT	GENERATED ALWAYS AS ENTITY ID,
PRIMARY KEY(conduit_entity_id)
        );
                            CREATE TABLE rel_outerStore_and_outerStore_inners (
                                left INT,
                                right INT,
                                constraint fk_outerStore_inners_right
                                    FOREIGN KEY(right)
                                        REFERENCES outerStore_inners(conduit_entity_id)
                                constraint fk_outerStore_left
                                    FOREIGN KEY(left)
                                        REFERENCES outerStore(conduit_entity_id)
                            )
                        
                        
                        insert into cities(name, location)
                        values ('detroit', 12)",
    "name": ".deploy/postgres/startup/init.sql",
  },
]
`;

exports[`struct containing struct stored: main files 1`] = `
Array [
  Object {
    "content": "
                            #![allow(non_snake_case)]
                            #![allow(non_camel_case_types)]
                            #![allow(redundant_semicolon)]
                            use tokio_postgres::{NoTls, Client};
                            use actix_web::{web, App, HttpResponse, HttpServer, Responder};
                            use std::env;
                            use serde::{Deserialize, Serialize};
                            use tokio_postgres::error::{Error};
                
                
                            struct AppData {
                                client: Client
                            }
                            
                            #[derive(Serialize, Deserialize)]
                            struct City {
                                name: String,
                                location: i32
                            }
                            
                    
                            
                        #[derive(Serialize, Deserialize)]
                        struct inner {
                            fieldA: f64
                        }
                    

                        #[derive(Serialize, Deserialize)]
                        struct outer {
                            inner: inner
                        }
                    
                    
                            
                    
                            #[actix_rt::main]
                            async fn main() -> std::io::Result<()> {
                                HttpServer::new(|| {
                                    App::new()
                                        .data_factory(|| make_app_data())
                                        .route(\\"/\\", web::get().to(index))
                                        
                                })
                                .bind(\\"0.0.0.0:8080\\")?
                                .run()
                                .await
                            }
                    
                            async fn index(data: web::Data<AppData>) -> impl Responder {
                                let mut rows = match data.client.query(\\"select name, location from cities\\", &[]).await {
                                    Ok(rows) => rows,
                                    Err(err) => panic!(\\"didn't succeed: {}\\", err)
                                };
                            
                                let mut out = Vec::with_capacity(rows.len());
                            
                                while let Some(row) = rows.pop() {
                                    out.push(City {
                                        name: row.get(0),
                                        location: row.get(1)
                                    })
                                }
                                return HttpResponse::Ok().json(out);
                            }
                
                            async fn make_app_data() -> Result<AppData, ()> {
                                let host = match env::var(\\"POSTGRES_SERVICE_HOST\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres location: {}\\", e)
                                };
                                let pwd = match env::var(\\"POSTGRES_PASSWORD\\") {
                                    Ok(pgloc) => pgloc,
                                    Err(e) => panic!(\\"didn't receive postgres password: {}\\", e)
                                };
                            
                                let (client, connection) = match tokio_postgres::connect(&format!(\\"host={} user=postgres password={}\\", host, pwd), NoTls).await {
                                    Ok(out) => out,
                                    Err(e) => panic!(\\"couldn't create connection: {}\\", e)
                                };
                                
                                // The connection object performs the actual communication with the database,
                                // so spawn it off to run on its own.
                                actix_rt::spawn(async move {
                                    if let Err(e) = connection.await {
                                        eprintln!(\\"connection error: {}\\", e);
                                    }
                                });
                                
                                
                                return Ok(AppData {
                                    client: client,
                                });
                            }
                            
                        ",
    "name": ".deploy/main/src/main.rs",
  },
]
`;

exports[`struct containing struct stored: postgres files 1`] = `
Array [
  Object {
    "content": "

                        CREATE TABLE cities (
                            name            varchar(80),
                            location        int
                        );
                
                        
        CREATE TABLE outerStore_inner (
            fieldA	double precision,
conduit_entity_id	INT	GENERATED ALWAYS AS ENTITY ID,
PRIMARY KEY(conduit_entity_id)
        );

        CREATE TABLE outerStore (
            inner	INT,
conduit_entity_id	INT	GENERATED ALWAYS AS ENTITY ID,
constraint fk_inner
                                FOREIGN KEY(inner)
                                    REFERENCES outerStore_inner(conduit_entity_id),
PRIMARY KEY(conduit_entity_id)
        );
                        
                        
                        insert into cities(name, location)
                        values ('detroit', 12)",
    "name": ".deploy/postgres/startup/init.sql",
  },
]
`;
